doc/mkepmlist.1
.\"
.\" Manual page for the ESP Package Manager (EPM).
.\"
.\" Copyright 2020 by Jim Jagielski
.\" Copyright 1999-2020 by Michael R Sweet
.\" Copyright 1999-2007 by Easy Software Products, all rights reserved.
.\"
.\" Licensed under the Apache License, Version 2.0 (the "License");
.\" you may not use this file except in compliance with the License.
.\" You may obtain a copy of the License at
.\"
.\"    http://www.apache.org/licenses/LICENSE-2.0
.\" 
.\" Unless required by applicable law or agreed to in writing, software
.\" distributed under the License is distributed on an "AS IS" BASIS,
.\" WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.\" See the License for the specific language governing permissions and
.\" limitations under the License.
.\"
.TH mkepmlist 1 "ESP Package Manager" "20 November 2020" "Jim Jagielski"
.SH NAME
mkepmlist \- make an epm list file from a directory.
.SH SYNOPSIS
.B mkepmlist
[
.B \-g
.I group
] [
.B \-u
.I user
] [
.B \-\-prefix
.I directory
]
.I directory
[ ...
.I directory
]
.SH DESCRIPTION
.B mkepmlist (1)
recursively generates file list entries for files, links, and directories.
The file list is send to the standard output.
.SH OPTIONS
.B mkepmlist
supports the following options:
.TP 5
\fB\-g \fIgroup\fR
Overrides the group ownership of the files in the specified directories with the specified group name.
.TP 5
\fB\-u \fIuser\fR
Overrides the user ownership of the files in the specified directories with the specified user name.
.TP 5
\fB\-\-prefix \fIdirectory\fR
Adds the specified directory to the destination path.
For example, if you installed files to "/opt/foo" and wanted to build a distribution that installed the files in "/usr/local", the following command would generate a file
list that is installed in "/usr/local":
.nf
.br
     mkepmlist \-\-prefix=/usr/local /opt/foo >foo.list
.fi
.SH SEE ALSO
.BR epm (1),
.BR epminstall (1),
.BR epm.list (5).
.SH COPYRIGHT
Copyright \[co] 1999-2020 by Michael R Sweet, All Rights Reserved.
Copyright \[co] 2020 by Jim Jagielski, All Rights Reserved.
.LP
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
.LP
   http://www.apache.org/licenses/LICENSE-2.0
.LP
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGES.md
Changes in EPM
==============

Changes in EPM 5.0.0
--------------------

- Enabling the GUI correctly checks for FLTK
- The Apache OpenOffice specific changes are no longer configure-time
  (buildtime) enabled, but run time, via the --aoo-mode CLI option.
- Relicensed to the ALv2 (via copyright holder)

Changes in EPM 4.5.2
--------------------

- Use `fakeroot` if available when building DEB packages to ensure
  correct file permissions
- Support `rpmbuild` for RHEL7 and similar platforms
- Brought back support for platforms deprecated in 4.5.0

Changes in EPM 4.5.1
--------------------

- Fixed Debian dependencies for min/max version (Issue #64)
- Fixed a typo in the epm.list file (Issue #78)
- Fixed macro usage in the setup.types man page (Issue #81)
- Fixed @INSTALL@ macro in makefile (Issue #84)
- Fixed Debian init script support (Issue #85)


Changes in EPM 4.5
------------------

- "make install" failed due to the README filename changing (Issue #59)
- The `mkepmlist` utility did not correctly handle filenames containing the
  `$` character (Issue #62)
- The configure script incorrectly substituted "NONE" for the installation
  prefix (Issue #67)
- Fixed some over-zealous permissions on temporary build directories (Issue #71)
- Fixed the mantohtml helper program - was depending on private CUPS headers for
  some reason (Issue #72)
- Fixed some build system issues (Issue #75, Issue #76)
- Documentation updates (Issue #74, Issue #77)
- Cleaned up old sprintf and strcpy usage in many places.

Changes in EPM 4.4.2
--------------------

- Support Apache OpenOffice patches (at configure time)

Changes in EPM 4.4.1
--------------------

- "make install" failed due to the README filename changing (Issue #59)
- The `mkepmlist` utility did not correctly handle filenames containing the
  `$` character (Issue #62)
- The configure script incorrectly substituted "NONE" for the installation
  prefix (Issue #67)
- Cleaned up old sprintf and strcpy usage in many places.


Changes in EPM 4.4
------------------

- The default prefix is now the usual `/usr/local` (Issue #45)
- Really fix 64-bit Intel packages on Debian-based OS's (Issue #48)
- Fixed a build issue on Solaris 11 (Issue #50)
- Fixed a bug in temporary file cleanup when symlinks are used (Issue #51)
- Added DESTDIR support to makefiles (Issue #55)
- Fixed RPM support on AIX (Issue #56)
- Reverted the hard links optimization from EPM 4.2 since it is causing
  problems with the latest version of RPM (Issue #57)
- Packages on macOS now use "macos" as the operating system name for
  consistency.


Changes in EPM 4.3
------------------

- Now use pkgbuild on newer versions of macOS, and added support for signed
  packages (Bug #497)
- Fixed some file handling issues when creating RPM packages (Bug #523)
- EPM now maps the x86_64 architecture to amd64 when creating Debian packages
  (Bug #295)
- %format stopped working in EPM 4.2 (Bug #296)
- %literal(spec) did not insert the literal content in the correct location
  (Bug #302)
- Fixed some incorrect string handling (Bug #290)
- Fixed a compatibility issue with RPM 4.8 (Bug #292)
- Fixed a build dependency problem (Bug #291)
- The EPM makefile now uses CPPFLAGS from the configure script (Bug #300)
- Updated the standard path used by portable package scripts to include
  /usr/gnu/bin for Solaris (Bug #301)
- Added support for %literal(control) in Debian packages (Bug #297)


Changes in EPM 4.2
------------------

- EPM now supports a %arch conditional directive (STR #27)
- EPM now uses hard links whenever possible instead of copying files for
  distribution (STR #21)
- EPM no longer puts files in /export in the root file set for AIX packages
  (STR #15)
- EPM did not work with newer versions of RPM (STR #23, STR #25)
- EPM did not clean up temporary files from Solaris packages (STR #20)
- Building Solaris gzip'd packages failed if the pkg.gz file already existed
  (STR #16)
- Fixed handling of %preremove and %postremove for AIX packages (STR #22)
- Fixed directory permissions in HP-UX packages (STR #24)
- Removed unnecessary quoting of "!" in filenames (STR #26)
- Added support for signed RPM packages (STR #19)
- Added support for inclusion of format-specific packaging files and directives
  via a %literal directive (STR #5)
- *BSD init scripts were not installed properly.
- EPM now displays a warning message when a variable is undefined (STR #10)
- *BSD dependencies on versioned packages are now specified correctly (STR #4)
- EPM now uses /usr/sbin/pkg_create on FreeBSD (STR #2)
- FreeBSD packages are now created with a .tbz extension (STR #1)
- FreeBSD packages incorrectly assumed that chown was installed in /bin (STR #3)
- Added support for an "lsb" package format which uses RPM with the LSB
  dependencies (STR #7)
- The configure script now supports a --with-archflags and no longer
  automatically builds universal binaries on macOS.
- The epm program now automatically detects when the setup GUI is not available,
  displays a warning message, and then creates a non-GUI package.
- RPM packages did not map %replaces to Obsoletes:


Changes in EPM 4.1
------------------

- macOS portable packages did not create a correct Uninstall application.
- The temporary package files for portable packages are now removed after
  creation of the .tar.gz file unless the -k (keep files) option is used.
- The RPM summary string for subpackages did not contain the first line of the
  package description as for other package formats.
- The setup and uninst GUIs now support installing and removing RPM packages.
- The setup GUI now confirms acceptance of all licenses prior to installing the
  first package.
- Subpackages are no longer automatically dependent on the main package.
- Multi-line descriptions were not embedded properly into portable package
  install/patch/remove scripts.
- Updated the setup and uninstall GUIs for a nicer look-n-feel.

mkepmlist.c
/*
 * List file generation utility for the ESP Package Manager (EPM).
 *
 * Copyright © 2020 by Jim Jagielski
 * Copyright 2003-2019 by Michael R Sweet
 * Copyright 2003-2005 by Easy Software Products
 * Copyright 2003 Andreas Voegele
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "epm.h"

/*
 * Lookup hash table structure...
 */

#define HASH_M 101

struct node {
    unsigned id; /* User or group ID */
    char *name;  /* User or group name */
};

/*
 * Globals...
 */

char *DefaultUser = NULL,   /* Default user for entries */
    *DefaultGroup = NULL;   /* Default group for entries */
struct node Users[HASH_M];  /* Hash table for users */
struct node Groups[HASH_M]; /* Hash table for groups */

/*
 * Functions...
 */

char *get_group(gid_t gid);
char *get_user(uid_t uid);
void hash_deinit(struct node *a);
unsigned hash_id(unsigned id);
void hash_init(struct node *a);
char *hash_insert(struct node *a, unsigned id, const char *name);
char *hash_search(struct node *a, unsigned id);
void info(void);
int process_dir(const char *srcpath, const char *dstpath);
int process_file(const char *src, const char *dstpath);
char *quote_string(char *q, const char *s, size_t qsize);
void usage(void);

/*
 * 'main()' - Scan directories to produce a distribution list.
 */

int                /* O - Exit status */
main(int argc,     /* I - Number of command-line arguments */
     char *argv[]) /* I - Command-line arguments */
{
    int i;              /* Looping var */
    const char *prefix, /* Installation prefix */
        *dstpath;       /* Destination path  */
    char dst[1024],     /* Destination */
        *ptr;           /* Pointer into filename */
    struct stat info;   /* File information */

    /*
     * Initialize the hash tables...
     */

    hash_init(Users);
    hash_init(Groups);

    /*
     * Loop through the command-line arguments, processing directories as
     * needed...
     */

    prefix = NULL;

    for (i = 1; i < argc; i++)
        if (strcmp(argv[i], "-u") == 0) {
            /*
             * -u username
             */

            i++;

            if (i >= argc)
                usage();

            DefaultUser = argv[i];
        } else if (strcmp(argv[i], "-g") == 0) {
            /*
             * -g groupname
             */

            i++;

            if (i >= argc)
                usage();

            DefaultGroup = argv[i];
        } else if (strcmp(argv[i], "--prefix") == 0) {
            i++;

            if (i >= argc)
                usage();

            prefix = argv[i];
        } else if (argv[i][0] == '-') {
            /*
             * Unknown option...
             */

            fprintf(stderr, "Unknown option \"%s\"!\n", argv[i]);
            usage();
        } else {
            /*
             * Directory name...
             */

            if (prefix)
                dstpath = prefix;
            else {
                if (argv[i][0] == '/')
                    dstpath = argv[i];
                else {
                    if (argv[i][0] == '.') {
                        for (dstpath = argv[i]; *dstpath == '.';) {
                            if (strncmp(dstpath, "./", 2) == 0)
                                dstpath += 2;
                            else if (strncmp(dstpath, "../", 3) == 0)
                                dstpath += 3;
                            else
                                break;
                        }

                        if (strcmp(dstpath, ".") == 0)
                            dstpath = "";
                    } else
                        dstpath = argv[i];

                    if (dstpath[0]) {
                        snprintf(dst, sizeof(dst), "/%s", dstpath);
                        dstpath = dst;
                    }
                }
            }

            if (!lstat(argv[i], &info)) {
                if (!S_ISDIR(info.st_mode)) {
                    if (!prefix) {
                        /*
                         * Remove trailing filename component from destination...
                         */

                        if (dstpath != dst) {
                            strlcpy(dst, dstpath, sizeof(dst));
                            dstpath = dst;
                        }

                        if ((ptr = strrchr(dstpath, '/')) != NULL)
                            *ptr = '\0';
                    }

                    process_file(argv[i], dstpath);
                } else
                    process_dir(argv[i], dstpath);
            }
        }

    /*
     * Free any memory we have left allocated...
     */

    hash_deinit(Users);
    hash_deinit(Groups);

    return (0);
}

/*
 * 'get_group()' - Get a group name for the given group ID.
 */

char *               /* O - Name of group or NULL */
get_group(gid_t gid) /* I - Group ID */
{
    char *name;       /* Name of group */
    struct group *gr; /* Group entry for group */
    char buf[16];     /* Temporary string buffer for group numbers */

    /*
     * Always return the default group if set...
     */

    if (DefaultGroup)
        return (DefaultGroup);

    /*
     * See if the group ID is already in the hash table...
     */

    if ((name = hash_search(Groups, gid)) != NULL)
        return (name);

    /*
     * Lookup the group ID in the password file...
     */

    setgrent();

    if ((gr = getgrgid(gid)) != NULL)
        name = gr->gr_name;
    else {
        /*
         * Unable to find group ID; just return the number...
         */

        snprintf(buf, sizeof(buf), "%u", (unsigned)gid);
        name = buf;
    }

    /*
     * Add the name to the hash table and return it...
     */

    return (hash_insert(Groups, gid, name));
}

/*
 * 'get_user()' - Get a user name for the given user ID.
 */

char *              /* O - Name of user or NULL */
get_user(uid_t uid) /* I - User ID */
{
    char *name;        /* Name of user */
    struct passwd *pw; /* Password entry for user */
    char buf[16];      /* Temporary string buffer for user numbers */

    /*
     * Always return the default user if set...
     */

    if (DefaultUser)
        return (DefaultUser);

    /*
     * See if the user ID is already in the hash table...
     */

    if ((name = hash_search(Users, uid)) != NULL)
        return (name);

    /*
     * Lookup the user ID in the password file...
     */

    setpwent();

    if ((pw = getpwuid(uid)) != NULL)
        name = pw->pw_name;
    else {
        /*
         * Unable to find user ID; just return the number...
         */

        snprintf(buf, sizeof(buf), "%u", (unsigned)uid);
        name = buf;
    }

    /*
     * Add the name to the hash table and return it...
     */

    return (hash_insert(Users, uid, name));
}

/*
 * 'hash_deinit()' - Deinitialize a hash table.
 */

void hash_deinit(struct node *a) /* I - Hash table */
{
    unsigned h; /* Looping var */

    for (h = 0; h < HASH_M; h++)
        if (a[h].name) {
            free(a[h].name);
            memset(a, 0, sizeof(struct node));
        }
}

/*
 * 'hash_id()' - Generate a hash for a user or group ID.
 */

unsigned             /* O - Hash number */
hash_id(unsigned id) /* I - User or group ID */
{
    return (id % HASH_M);
}

/*
 * 'hash_init()' - Initialize a hash table.
 */

void hash_init(struct node *a) /* I - Hash table to initialize */
{
    /*
     * Zero out the hash table...
     */
    memset(a, 0, HASH_M * sizeof(struct node));
}

/*
 * 'hash_insert()' - Insert a new entry in a hash table.
 */

char *                        /* O - New user or group name or NULL */
hash_insert(struct node *a,   /* I - Hash table */
            unsigned id,      /* I - User or group ID */
            const char *name) /* I - User or group name */
{
    unsigned h, /* Current hash value */
        start;  /* Starting hash value */

    for (h = start = hash_id(id); a[h].name;) {
        h = (h + 1) % HASH_M;
        if (h == start)
            return (NULL);
    }

    a[h].id = id;
    a[h].name = strdup(name);

    return (a[h].name);
}

/*
 * 'hash_search()' - Find a user or group ID in the hash table...
 */

char *                      /* O - User or group name, or NULL */
hash_search(struct node *a, /* I - Hash table */
            unsigned id)    /* I - User or group ID */
{
    unsigned h, /* Current hash value */
        start;  /* Starting hash value */

    for (h = start = hash_id(id); a[h].name;) {
        h = (h + 1) % HASH_M;

        if (a[h].id == id)
            return (a[h].name);

        if (h == start)
            break;
    }

    return (NULL);
}

/*
 * 'info()' - Show the EPM copyright and license.
 */

void info(void) {
    puts(EPM_VERSION);
    puts("Copyright 1999-2019 by Michael R Sweet.");
    puts("Copyright 2020 by Jim Jagielski.");
    puts("");
    puts("EPM is free software and comes with ABSOLUTELY NO WARRANTY; for details");
    puts("see the Apache License version 2.0 in the file LICENSE or at");
    puts("\"https://www.apache.org/licenses/LICENSE-2.0\".  Report all problems to");
    puts("\"https://github.com/jimjag/epm/issues\".");
    puts("");
}

/*
 * 'process_dir()' - Process a directory...
 */

int                              /* O - 0 on success, -1 on error */
process_dir(const char *srcpath, /* I - Source path */
            const char *dstpath) /* I - Destination path */
{
    DIR *dir;            /* Directory to read from */
    struct dirent *dent; /* Current directory entry */
    size_t srclen;       /* Length of source path */
    char src[1024];      /* Temporary source path */

    /*
     * Try opening the source directory...
     */

    if ((dir = opendir(srcpath)) == NULL) {
        fprintf(stderr, "mkepmlist: Unable to open directory \"%s\": %s.\n", srcpath,
                strerror(errno));

        return (-1);
    }

    /*
     * Read from the directory...
     */

    srclen = strlen(srcpath);

    while ((dent = readdir(dir)) != NULL) {
        /*
         * Skip "." and ".."...
         */

        if (strcmp(dent->d_name, ".") == 0 || strcmp(dent->d_name, "..") == 0)
            continue;

        /*
         * Process file...
         */

        if (srclen > 0 && srcpath[srclen - 1] == '/')
            snprintf(src, sizeof(src), "%s%s", srcpath, dent->d_name);
        else
            snprintf(src, sizeof(src), "%s/%s", srcpath, dent->d_name);

        if (process_file(src, dstpath)) {
            closedir(dir);
            return (-1);
        }
    }

    closedir(dir);
    return (0);
}

/*
 * 'process_file()' - Process a file...
 */

int                               /* O - 0 on success, -1 on error */
process_file(const char *src,     /* I - Source path */
             const char *dstpath) /* I - Destination path */
{
    const char *srcptr;  /* Pointer into source path */
    struct stat srcinfo; /* Information on the source file */
    ssize_t linklen;     /* Length of link path */
    size_t dstlen;       /* Length of destination path */
    char link[1024],     /* Link for source */
        dst[1024],       /* Temporary destination path */
        qdst[1024],      /* Quoted destination */
        qsrc[1024];      /* Quoted source/link */

    /*
     * Get source file info...
     */

    dstlen = strlen(dstpath);

    if ((srcptr = strrchr(src, '/')) != NULL)
        srcptr++;
    else
        srcptr = src;

    if (dstlen > 0 && dstpath[dstlen - 1] == '/')
        snprintf(dst, sizeof(dst), "%s%s", dstpath, srcptr);
    else
        snprintf(dst, sizeof(dst), "%s/%s", dstpath, srcptr);

    if (lstat(src, &srcinfo)) {
        fprintf(stderr, "mkepmlist: Unable to stat \"%s\": %s.\n", src, strerror(errno));
        return (-1);
    }

    /*
     * Process accordingly...
     */

    if (S_ISDIR(srcinfo.st_mode)) {
        /*
         * Directory...
         */

        printf("d %o %s %s %s -\n", (unsigned)(srcinfo.st_mode & 07777),
               get_user(srcinfo.st_uid), get_group(srcinfo.st_gid),
               quote_string(qdst, dst, sizeof(qdst)));

        if (process_dir(src, dst))
            return (-1);
    } else if (S_ISLNK(srcinfo.st_mode)) {
        /*
         * Symlink...
         */

        if ((linklen = readlink(src, link, sizeof(link) - 1)) < 0) {
            fprintf(stderr, "mkepmlist: Unable to read symlink \"%s\": %s.\n", src,
                    strerror(errno));
            return (-1);
        }

        link[linklen] = '\0';

        printf("l %o %s %s %s %s\n", (unsigned)(srcinfo.st_mode & 07777),
               get_user(srcinfo.st_uid), get_group(srcinfo.st_gid),
               quote_string(qdst, dst, sizeof(qdst)),
               quote_string(qsrc, link, sizeof(qsrc)));
    } else if (S_ISREG(srcinfo.st_mode)) {
        /*
         * Regular file...
         */

        printf("f %o %s %s %s %s\n", (unsigned)(srcinfo.st_mode & 07777),
               get_user(srcinfo.st_uid), get_group(srcinfo.st_gid),
               quote_string(qdst, dst, sizeof(qdst)),
               quote_string(qsrc, src, sizeof(qsrc)));
    }

    return (0);
}

/*
 * 'quote_string()' - Quote space, backslash, and $ in a string...
 */

char *                      /* O - Quoted string */
quote_string(char *q,       /* I - Quoted string buffer */
             const char *s, /* I - Original string */
             size_t qsize)  /* I - Size of buffer */
{
    char *qptr, *qend;

    for (qptr = q, qend = q + qsize - 2; *s && qptr < qend;) {
        if (*s == '\\' || *s == ' ' || *s == '\t')
            *qptr++ = '\\';
        else if (*s == '$')
            *qptr++ = '$';

        *qptr++ = *s++;
    }

    *qptr = '\0';

    return (q);
}

/*
 * 'usage()' - Show command-line usage instructions.
 */

void usage(void) {
    info();

    puts("Usage: mkepmlist [options] directory [... directory] >filename.list");
    puts("Options:");
    puts("-g group              Set group name for files.");
    puts("-u user               Set user name for files.");
    puts("--prefix directory    Set directory prefix for files.");

    exit(1);
}
